liblsp/Server.cpp

```diff
@@ -0,0 +1,391 @@
+
+	Trace trace = Trace::Off; // The initial trace setting. If omitted trace is disabled ('off').  // TODO: ^^ should be overridable by the solls CLI param
+	if (Json::Value value = _args["trace"]; value)
+	{
+		string const name = value.asString();
+		if (name == "messages")
+			trace = Trace::Messages;
+		else if (name == "verbose")
+			trace = Trace::Verbose;
+		else if (name == "off")
+			trace = Trace::Off;
+	}
+
+	std::vector<WorkspaceFolder> workspaceFolders; // initial configured workspace folders
+	if (Json::Value folders = _args["workspaceFolders"]; folders)
+	{
+		for (Json::Value folder: folders)
+		{
+			WorkspaceFolder wsFolder{};
+			wsFolder.name = folder["name"].asString();
+			wsFolder.uri = folder["uri"].asString();
+			workspaceFolders.emplace_back(move(wsFolder));
+		}
+	}
+	// TODO: initializationOptions
+	map<string, string> settings{};
+
+	// TODO: ClientCapabilities
+	// ...
+
+	auto const info = initialize(move(rootUri), move(settings), trace, move(workspaceFolders));
+
+	// {{{ encoding
+	Json::Value jsonReply;
+
+	if (!info.serverName.empty())
+	{
+		jsonReply["serverInfo"]["name"] = info.serverName;
+		if (!info.serverVersion.empty())
+			jsonReply["serverInfo"]["version"] = info.serverVersion;
+	}
+
+	jsonReply["hoverProvider"] = info.supportsHover;
+	jsonReply["capabilities"]["hoverProvider"] = info.supportsHover;
+	jsonReply["capabilities"]["textDocumentSync"]["openClose"] = true;
+	jsonReply["capabilities"]["textDocumentSync"]["change"] = true;
+	jsonReply["capabilities"]["definitionProvider"] = info.supportsDefinition;
+	jsonReply["capabilities"]["documentHighlightProvider"] = info.supportsDocumentHighlight;
+	jsonReply["capabilities"]["referencesProvider"] = info.supportsReferences;
+
+	m_client.reply(_id, jsonReply);
+	// }}}
+}
+
+void Server::handle_initialized(MessageId /*_id*/, Json::Value const& /*_args*/)
+{
+	// nothing to decode
+	initialized();
+	// nothing to encode
+}
+
+void Server::handle_shutdown(MessageId /*_id*/, Json::Value const& /*_args*/)
+{
+	logInfo("Shutdown requested");
+	m_shutdownRequested = true;
+}
+
+void Server::handle_exit(MessageId _id, Json::Value const& /*_args*/)
+{
+	m_exitRequested = true;
+	auto const exitCode = m_shutdownRequested ? 0 : 1;
+
+	Json::Value jsonReply = Json::intValue;
+	jsonReply = exitCode;
+
+	m_client.reply(_id, jsonReply);
+}
+
+void Server::handle_cancelRequest(MessageId, Json::Value const&)
+{
+	// for now we don't do anything because we're synchronous.
+}
+
+void Server::handle_textDocument_didOpen(MessageId /*_id*/, Json::Value const& _args)
+{
+	// decoding
+	if (!_args["textDocument"])
+		return;
+
+	auto const uri = _args["textDocument"]["uri"].asString();
+	auto const languageId = _args["textDocument"]["languageId"].asString();
+	auto const version = _args["textDocument"]["version"].asInt();
+	auto const text = _args["textDocument"]["text"].asString();
+
+	documentOpened(uri, languageId, version, text);
```

Comment:
Why is this not inlined?
Is it really useful to have the distinction between Server and LanguageServer if most of the function just "forward"?